pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/map/ImmutableMapDataSource.java

```diff
@@ -0,0 +1,157 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.local.segment.index.map;
+
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.apache.pinot.segment.spi.ColumnMetadata;
+import org.apache.pinot.segment.spi.datasource.DataSourceMetadata;
+import org.apache.pinot.segment.spi.index.IndexReader;
+import org.apache.pinot.segment.spi.index.column.ColumnIndexContainer;
+import org.apache.pinot.segment.spi.index.reader.ForwardIndexReader;
+import org.apache.pinot.segment.spi.index.reader.ForwardIndexReaderContext;
+import org.apache.pinot.segment.spi.index.reader.MapIndexReader;
+import org.apache.pinot.segment.spi.partition.PartitionFunction;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+@SuppressWarnings("rawtypes")
+public class ImmutableMapDataSource extends BaseMapDataSource {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ImmutableMapDataSource.class);
+  private final MapIndexReader _mapIndexReader;
+
+  public ImmutableMapDataSource(ColumnMetadata columnMetadata, ColumnIndexContainer columnIndexContainer) {
+    super(new ImmutableMapDataSourceMetadata(columnMetadata), columnIndexContainer);
+    MapIndexReader mapIndexReader = getMapIndex();
+    if (mapIndexReader == null) {
+      // Fallback to use forward index
+      ForwardIndexReader<?> forwardIndex = getForwardIndex();
+      if (forwardIndex instanceof MapIndexReader) {
+        mapIndexReader = (MapIndexReader) forwardIndex;
+      } else {
+        mapIndexReader = new MapIndexReaderWrapper(forwardIndex, getFieldSpec());
+      }
+    }
+    _mapIndexReader = mapIndexReader;
+  }
+
+  @Override
+  public MapIndexReader<ForwardIndexReaderContext, IndexReader> getMapIndexReader() {
+    return _mapIndexReader;
+  }
+
+  @Override
+  public DataSourceMetadata getKeyDataSourceMetadata(String key) {
+    return null;
+  }
+
+  @Override
+  public ColumnIndexContainer getKeyIndexContainer(String key) {
+    return null;
+  }
+
+  private static class ImmutableMapDataSourceMetadata implements DataSourceMetadata {
+    final FieldSpec _fieldSpec;
+    final int _numDocs;
+    final int _numValues;
+    final int _maxNumValuesPerMVEntry;
+    final int _cardinality;
+    final PartitionFunction _partitionFunction;
+    final Set<Integer> _partitions;
+    final Comparable _minValue;
+    final Comparable _maxValue;
+
+    ImmutableMapDataSourceMetadata(ColumnMetadata columnMetadata) {
+      _fieldSpec = columnMetadata.getFieldSpec();
+      _numDocs = columnMetadata.getTotalDocs();
+      _numValues = columnMetadata.getTotalNumberOfEntries();
+      if (_fieldSpec.isSingleValueField()) {
+        _maxNumValuesPerMVEntry = -1;
+      } else {
+        _maxNumValuesPerMVEntry = columnMetadata.getMaxNumberOfMultiValues();
```

Comment:
MultiValue MAP Column is not supported. What does this mean?

Reply:
MAP is considered as a single value column.